guides/recipes/anti-patterns.md

```diff
@@ -0,0 +1,843 @@
+- Resource contention between quick and long-running jobs
+- Priority inversion (critical jobs waiting behind non-critical ones)
+- Poor overall system throughput
+- Difficulty in monitoring and troubleshooting
+
+### Example
+
+```elixir
+defmodule MyApp.Workers.QuickEmailNotification do
+  # Missing queue definition - will use :default
+  use Oban.Worker
+
+  @impl Oban.Worker
+  def perform(%{args: %{"user_id" => user_id, "message" => message}}) do
+    # Quick operation that shouldn't be blocked
+    MyApp.Notifications.email_user(user_id, message)
+  end
+end
+
+defmodule MyApp.Workers.LongRunningReportGenerator do
+  # Also using default queue - will block other jobs
+  use Oban.Worker
+
+  @impl Oban.Worker
+  def perform(%{args: %{"report_id" => report_id}}) do
+    # Long-running operation that could block other jobs for minutes
+    MyApp.Reports.generate_report(report_id)
+  end
+end
+```
+
+### Refactoring
+
+Create a thoughtful queue structure based on job characteristics, leveraging Oban Pro's Smart Engine for dynamic concurrency:
+
+```elixir
+# In configuration
+config :my_app, Oban,
+  engine: Oban.Pro.Engines.Smart,
+  repo: MyApp.Repo,
+  plugins: [...],
+  queues: [
+    default: 10,         # General purpose queue
+    notifications: 20,    # High concurrency for quick operations
+    reports: 2,           # Limited concurrency for resource-intensive operations
+    external_api: 5       # Controlled concurrency for external API calls
+  ]
+```
+
+Assign jobs to appropriate queues:
+
+```elixir
+defmodule MyApp.Workers.QuickEmailNotification do
+  use Oban.Worker,
+    queue: :notifications,  # High-throughput queue
+    priority: 1             # Higher priority (lower number)
+
+  @impl Oban.Worker
+  def perform(%{args: %{"user_id" => user_id, "message" => message}}) do
+    MyApp.Notifications.email_user(user_id, message)
+  end
+end
+
+defmodule MyApp.Workers.LongRunningReportGenerator do
+  use Oban.Worker,
+    queue: :reports,  # Limited concurrency queue
+    priority: 3       # Lower priority (higher number)
+
+  @impl Oban.Worker
+  def perform(%{args: %{"report_id" => report_id}}) do
+    MyApp.Reports.generate_report(report_id)
+  end
+end
+```
+
+### Additional Remarks
+
+Design your queue structure based on these job characteristics:
+
+| Characteristic | Considerations |
+|----------------|---------------|
+| **Duration** | Separate quick jobs from long-running ones |
+| **Resource usage** | Limit concurrency for resource-intensive operations |
+| **Priority** | Use both queues and priority values for fine-grained control |
+| **External dependencies** | Create dedicated queues for rate-limited external services |
+| **Business importance** | Ensure critical operations get processed first |
+
+The Smart Engine will dynamically adjust concurrency based on job execution patterns, but it works best when you start with thoughtfully designed queue structure.
+
+## Anti-Pattern 7: Insufficient Error Handling in Jobs
```

Comment:
This is a new one, not in the issue. It arose from a discussion within my team where it was not clear the distinction on when to use `{:discard, any()}` and when to use `{:cancel, any()}`

My take on this is that `:discard` should be used either when attempts are exhausted or we faced a unrecoverable error. And `:cancel` should be used when although not successful, it's not an error – kinda like a `no-op` return

Reply:
Noted on Slack:

> Regarding the `cancel` vs `discard` discussion, the discard tuple is deprecated and undocumented now. Discard should always be because of exhausted retries, and cancel is always a manual action.